lib/publiccloud/utils.pm

```diff
@@ -20,19 +20,37 @@ use warnings;
 use testapi;
 use utils;
 use version_utils;
+use publiccloud::ssh_interactive;
 
 our @EXPORT = qw(select_host_console is_publiccloud is_byos is_ondemand is_ec2 is_azure is_gce);
 
-# Select console on the test host, regardless of the TUNNELED variable.
+# Select console on the test host, if force is set, the interactive session will
+# be destroyed. If called in TUNNELED environment, this function die.
+#
+# select_host_console(force => 1)
+#
 sub select_host_console {
-    my (@args) = @_;
-    if (check_var('TUNNELED', '1')) {
-        select_console('tunnel-console', @args);
-    } else {
-        select_console('root-console', @args);
+    my (%args) = @_;
+    $args{force} //= 0;
+    my $tunneled = get_var('TUNNELED');
+
+    if ($tunneled && check_var('_SSH_TUNNELS_INITIALIZED', 1)) {
+        die("Called select_host_console but we are in TUNNELED mode") unless ($args{force});
+
+        opensusebasetest::select_serial_terminal();
+        ssh_interactive_leave();
+
+        select_console('tunnel-console', await_console => 0);
+        send_key "ctrl-c";
+        send_key "ret";
+
+        set_var('_SSH_TUNNELS_INITIALIZED', 0);
+        sleep 3;
```

Comment:
You are using `sleep()` or are simulating a sleep-like functionality. This is introducing the risk of either unnecessarily wasting time or just shifting race conditions and making them even harder to investigate. You *can* use sleep but there are only very limited cases where is the best solution (e.g. yield threads in low-level system code). Here you have the classical problem in software design of [Synchronization](https://en.wikipedia.org/wiki/Synchronization_(computer_science)) . sleep should not be used in place of proper synchronization methods. https://blogs.msmvps.com/peterritchie/2007/04/26/thread-sleep-is-a-sign-of-a-poorly-designed-program/ is one of multiple references that explains this: "The original problem is likely a timing/synchronization issue, ignoring it by hiding it with Thread.Sleep is only going to delay the problem and make it occur in random, hard to reproduce ways." And if the problem does not occur at least time is wasted. https://testing.googleblog.com/2021/03/test-flakiness-one-of-main-challenges.html states that as well: "Do NOT add arbitrary delays as these can become flaky again over time and slow down the test unnecessarily."

We are waiting because some condition changes some time ... keyword(s) is/are *some time*. It is important to know the exact condition that the code should wait for and check for that instead of delaying program execution for an arbitrary amount of time. For openQA tests in general just try to do what you as a human user would also do. You would not look on your watch waiting for exactly 100s before you look at the screen again, right? :)

See https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/CONTRIBUTING.md#coding-style for more details